File: frontend/src/models/transcript.js
import { KeywordUsageModel } from './keyword-usage';

export class TranscriptModel {
  // Server Fields
  id;
  session_device_id;
  start_time;
  length;
  question;
  transcript;
  word_count;
  direction;
  emotional_tone_value;
  analytic_thinking_value;
  clout_value;
  authenticity_value;
  certainty_value;
  keywords;
  speaker_tag;

  static fromJson(json) {
    const model = new TranscriptModel();
    model.id = json['id'];
    model.session_device_id = json['session_device_id'];
    model.start_time = json['start_time'];
    model.length = json['length'];
    model.question = json['question'];
    model.transcript = json['transcript'];
    model.word_count = json['word_count'];
    model.direction = json['direction'];
    model.emotional_tone_value = json['emotional_tone_value'];
    model.analytic_thinking_value = json['analytic_thinking_value'];
    model.clout_value = json['clout_value'];
    model.authenticity_value = json['authenticity_value'];
    model.certainty_value = json['certainty_value'];
    model.speaker_tag = json['speaker_tag'];
    model.keywords = KeywordUsageModel.fromJsonList(json['keywords']);
    model.keywords.forEach(k => k.transcript_id = model.id);
    return model;
  }

  // Converts JSON to TranscriptModel[]
  static fromJsonList(jsonArray) {
    const transcripts = [];
    for (const el of jsonArray) {
      transcripts.push(TranscriptModel.fromJson(el));
    }
    return transcripts;
  }

  static tracker( index, transcript) {
    return transcript.id;
  }
}

File: server/tables/transcript.py
from app import db

class Transcript(db.Model):
    __tablename__ = 'transcript'
    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    session_device_id = db.Column(db.Integer, db.ForeignKey('session_device.id'), nullable=False)
    start_time = db.Column(db.Integer, nullable=False)
    length = db.Column(db.Integer, nullable=False)
    question = db.Column(db.Boolean, nullable=False)
    transcript = db.Column(db.Text, nullable=False)
    word_count = db.Column(db.Integer, nullable=False)
    direction = db.Column(db.Integer)
    emotional_tone_value = db.Column(db.Integer)
    analytic_thinking_value = db.Column(db.Integer)
    clout_value = db.Column(db.Integer)
    authenticity_value = db.Column(db.Integer)
    certainty_value = db.Column(db.Integer)
    topic_id = db.Column(db.Integer)
    speaker_tag = db.Column(db.String(64))
    speaker_id = db.Column(db.Integer)

    keywords = db.relationship("KeywordUsage", lazy='joined', uselist=True)
    metrics = db.relationship("SpeakerTranscriptMetrics", back_populates="transcript", cascade="all, delete",passive_deletes=True)

    def __hash__(self):
        return hash((self.id))

    def __init__(self, session_device_id, start_time, length, transcript, question, direction, emotional_tone, analytic_thinking, clout, authenticity, certainty, topic_id, speaker_tag, speaker_id):
        self.session_device_id = session_device_id
        self.start_time = start_time
        self.length = length
        self.end_time = start_time + length
        self.question = question
        self.transcript = transcript
        self.word_count = len(transcript.split())
        self.direction = direction
        self.emotional_tone_value = emotional_tone
        self.analytic_thinking_value = analytic_thinking
        self.clout_value = clout
        self.authenticity_value = authenticity
        self.certainty_value = certainty
        self.topic_id = topic_id
        self.speaker_tag = speaker_tag
        self.speaker_id = speaker_id


    def json(self):
        return dict(
            id=self.id,
            session_device_id=self.session_device_id,
            start_time=self.start_time,
            length=self.length,
            question=self.question,
            transcript=self.transcript,
            direction=self.direction,
            emotional_tone_value=self.emotional_tone_value,
            analytic_thinking_value=self.analytic_thinking_value,
            clout_value=self.clout_value,
            authenticity_value=self.authenticity_value,
            certainty_value=self.certainty_value,
            word_count=self.word_count,
            topic_id=self.topic_id,
            speaker_tag=self.speaker_tag,
            speaker_id=self.speaker_id,
            keywords=[keyword.json(suppress=['transcript_id']) for keyword in self.keywords]
        )

    @staticmethod
    def verify_fields(name=None):
        message = None
        # Not implemented
        return message == None, message
